sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala

```diff
@@ -139,6 +140,34 @@ case class InsertIntoHadoopFsRelationCommand(
     }
 
     if (doInsertion) {
+      // For dynamic partition overwrite, we do not delete partition directories ahead.
```

Comment:
I believe `InsertIntoHadoopFsRelationCommand` already deletes partition directories before reaching this point.

Reply:
> I believe `InsertIntoHadoopFsRelationCommand` already deletes partition directories before reaching this point.

Yes, it's a problem.  As mentioned in  https://github.com/apache/spark/pull/30057#discussion_r506791318.
Maybe we add a check before execute is best. But in that way we can't support dynamicPartitionFilter.